File: serializer-deserializer/src/main/java/com/amazonaws/services/schemaregistry/serializers/avro/AvroSerializer.java
/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates.
 * Licensed under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.amazonaws.services.schemaregistry.serializers.avro;

import com.amazonaws.services.schemaregistry.common.GlueSchemaRegistryDataFormatSerializer;
import com.amazonaws.services.schemaregistry.exception.AWSSchemaRegistryException;
import com.amazonaws.services.schemaregistry.utils.AVROUtils;
import com.amazonaws.services.schemaregistry.utils.AvroRecordType;
import com.google.common.annotations.VisibleForTesting;
import com.google.common.cache.CacheBuilder;
import com.google.common.cache.CacheLoader;
import com.google.common.cache.LoadingCache;
import lombok.AllArgsConstructor;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NonNull;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import org.apache.avro.Schema;
import org.apache.avro.generic.GenericData;
import org.apache.avro.generic.GenericRecord;
import org.apache.avro.io.BinaryEncoder;
import org.apache.avro.io.DatumWriter;
import org.apache.avro.io.EncoderFactory;
import org.apache.avro.specific.SpecificRecord;

import java.io.ByteArrayOutputStream;

/**
 * Avro serialization helper.
 */
@Slf4j
public class AvroSerializer implements GlueSchemaRegistryDataFormatSerializer {
    private AVROUtils avroUtils = AVROUtils.getInstance();
    private static final long MAX_DATUM_WRITER_CACHE_SIZE = 100;

    @NonNull
    @VisibleForTesting
    protected final LoadingCache<DatumWriterCacheKey, DatumWriter<Object>> datumWriterCache;

    public AvroSerializer() {
        this.datumWriterCache =
            CacheBuilder
                .newBuilder()
                .maximumSize(MAX_DATUM_WRITER_CACHE_SIZE)
                .build(new DatumWriterCache());
    }

    @Override
    public byte[] serialize(Object data) {
        byte[] bytes;
        bytes = serialize(data, createDatumWriter(data));

        return bytes;
    }

    /**
     * This method is used to create Avro datum writer for serialization. Based on
     * the Avro record type, GenericDatumWriter or SpecificDatumWriter will be
     * created.
     *
     * @param object the Avro message
     * @return Avro datum writer for serialization
     */
    private DatumWriter<Object> createDatumWriter(Object object) {
        org.apache.avro.Schema schema = AVROUtils.getInstance()
                .getSchema(object);
        if (object instanceof SpecificRecord) {
            return getSpecificDatumWriter(schema);
        } else if (object instanceof GenericRecord) {
            return getGenericDatumWriter(schema);
        } else if (object instanceof GenericData.EnumSymbol) {
            return getGenericDatumWriter(schema);
        } else if (object instanceof GenericData.Array) {
            return getGenericDatumWriter(schema);
        } else if (object instanceof GenericData.Fixed) {
            return getGenericDatumWriter(schema);
        } else {
            String message =
                String.format("Unsupported type passed for serialization: %s", object);
            throw new AWSSchemaRegistryException(message);
        }
    }

    @SneakyThrows
    private DatumWriter<Object> getSpecificDatumWriter(Schema schema) {
        DatumWriterCacheKey datumWriterCacheKey = new DatumWriterCacheKey(schema, AvroRecordType.SPECIFIC_RECORD);
        return datumWriterCache.get(datumWriterCacheKey);
    }

    @SneakyThrows
    private DatumWriter<Object> getGenericDatumWriter(Schema schema) {
        DatumWriterCacheKey datumWriterCacheKey = new DatumWriterCacheKey(schema, AvroRecordType.GENERIC_RECORD);
        return datumWriterCache.get(datumWriterCacheKey);
    }

    /**
     * Serialize the Avro message to bytes
     *
     * @param data   the Avro message for serialization
     * @param writer Avro DatumWriter for serialization
     * @return the serialized byte array
     * @throws AWSSchemaRegistryException AWS Schema Registry Exception
     */
    private byte[] serialize(Object data, DatumWriter<Object> writer) {
        return encodeData(data, writer);
    }

    private byte[] encodeData(Object object, DatumWriter<Object> writer) {
        ByteArrayOutputStream actualDataBytes = new ByteArrayOutputStream();
        BinaryEncoder encoder = EncoderFactory.get().directBinaryEncoder(actualDataBytes, null);
        try {
            writer.write(object, encoder);
            encoder.flush();
        } catch (Exception e) {
            throw new AWSSchemaRegistryException(e.getMessage(), e);
        }
        return actualDataBytes.toByteArray();
    }

    /**
     * Get the schema definition.
     *
     * @param object object for which schema definition has to be derived
     * @return schema string
     */
    @Override
    public String getSchemaDefinition(@NonNull Object object) {
        return avroUtils.getSchemaDefinition(object);
    }

    public void validate(Object data) {
        //No-op
        //Avro format assumes that the passed object contains schema and data that are mutually conformant.
        //We cannot validate the data against the schema.
    }

    public void validate(String schemaDefinition, byte[] data) {
        //No-op
        //We cannot determine accurately if the data bytes match the schema as Avro bytes don't contain the field names.
    }

    @AllArgsConstructor
    @Getter
    @EqualsAndHashCode
    private static class DatumWriterCacheKey {
        @NonNull
        private final Schema schema;
        @NonNull
        private final AvroRecordType avroRecordType;
    }

    private static class DatumWriterCache extends CacheLoader<DatumWriterCacheKey, DatumWriter<Object>> {
        @Override
        public DatumWriter<Object> load(DatumWriterCacheKey datumWriterCacheKey) {
            Schema schema = datumWriterCacheKey.getSchema();
            AvroRecordType avroRecordType = datumWriterCacheKey.getAvroRecordType();
            return DatumWriterInstance.get(schema, avroRecordType);
        }
    }
}

File: native-schema-registry/csharp/AWSGsrSerDe/AWSGsrSerDe/serializer/avro/AvroSerializer.cs
using System;
using System.Collections.Generic;
using System.IO;
using Avro;
using Avro.Generic;
using Avro.IO;
using Avro.Specific;
using AWSGsrSerDe.common;
using Microsoft.Extensions.Caching.Memory;

namespace AWSGsrSerDe.serializer.avro
{
    /// <summary>
    /// Avro serialization helper.
    /// </summary>
    public class AvroSerializer : IDataFormatSerializer
    {

        private readonly MemoryCache _datumWriterCache;

        /// <summary>
        /// Size of datum writer cache for testing purpose
        /// </summary>
        public int CacheSize => _datumWriterCache.Count;

        private AvroSerializer()
        {
            // TODO: make cache size limit configurable
            _datumWriterCache = new MemoryCache(new MemoryCacheOptions{SizeLimit = 1000});
        }

        /// <summary>
        ///  Constructor for Avro serializer that accepts configuration elements.
        /// </summary>
        /// <param name="configuration">configuration elements</param>
        public AvroSerializer(GlueSchemaRegistryConfiguration configuration = null)
        :this()
        {
        }

        /// <summary>
        /// Serialize the Avro message to bytes
        /// </summary>
        /// <param name="data">the Avro message for serialization</param>
        /// <returns>the serialized byte array</returns>
        public byte[] Serialize(object data)
        {
            // ReSharper disable once ConvertIfStatementToReturnStatement
            if (data == null)
                return null;
            return Serialize(data, GetOrCreateDatumWriter(data));
        }

        /// <summary>
        /// Get the schema definition.
        /// </summary>
        /// <param name="data">object for which schema definition has to be derived</param>
        /// <returns>schema string</returns>
        public string GetSchemaDefinition(object data)
        {
            var schema = GetSchema(data);
            return schema.ToString();
        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="schemaDefinition"></param>
        /// <param name="data"></param>
        public void Validate(string schemaDefinition, byte[] data)
        {
            //No-op
            //We cannot determine accurately if the data bytes match the schema as Avro bytes don't contain the field names.
        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="data"></param>
        public void Validate(object data)
        {
            //No-op
            //Avro format assumes that the passed object contains schema and data that are mutually conformant.
            //We cannot validate the data against the schema.
        }

        private DatumWriter<object> GetOrCreateDatumWriter(object data)
        {
            var schema = GetSchema(data);
            var dataType = data.GetType().ToString();
            var cacheKey = schema + dataType;
            var datumWriter = _datumWriterCache.GetOrCreate(
                cacheKey,
                entry =>
                {
                    entry.Size = 1;
                    return NewDatumWriter(data, schema);
                });
            return datumWriter;
        }
        
        private static byte[] Serialize(object data, DatumWriter<object> datumWriter)
        {
            var memoryStream = new MemoryStream();
            Encoder encoder = new BinaryEncoder(memoryStream);
            datumWriter.Write(data, encoder);
            return memoryStream.ToArray();
        }

        private static Schema GetSchema(object data)
        {
            return data switch
            {
                GenericRecord record => record.Schema,
                GenericEnum record => record.Schema,
                ISpecificRecord record => record.Schema,
                GenericFixed record => record.Schema,
                _ => throw new AwsSchemaRegistryException("Unsupported Avro Data formats")
            };
        }

        private static DatumWriter<object> NewDatumWriter(object data, Schema schema)
        {
            return data switch
            {
                ISpecificRecord _ => new SpecificDatumWriter<object>(schema),
                GenericRecord _ => new GenericWriter<object>(schema),
                GenericEnum _ => new GenericWriter<object>(schema),
                GenericFixed _=> new GenericWriter<object>(schema),
                _ => throw new AwsSchemaRegistryException($"Unsupported type passed for serialization: {data}")
            };
        }
    }
}
